src/renderer/vt/VtSequences.cpp

```diff
@@ -464,9 +463,18 @@ using namespace Microsoft::Console::Render;
 // - The hyperlink URI
 // Return Value:
 // - S_OK if we succeeded, else an appropriate HRESULT for failing to allocate or write.
-[[nodiscard]] HRESULT VtEngine::_SetHyperlink(std::wstring uri) noexcept
+[[nodiscard]] HRESULT VtEngine::_SetHyperlink(const std::wstring& uri, const USHORT& id) noexcept
 {
-    const auto convertedUri = ConvertToA(CP_UTF8, uri);
-    const auto hyperlinkFormat = "\x1b]8;;" + convertedUri + "\x7";
-    return _Write(hyperlinkFormat);
+    if (uri.empty())
+    {
+        // Closing OSC8 sequence
+        const auto hyperlinkFormat = (std::wstring)L"\x1b]8;;\x9c";
+        return _Write(til::u16u8(hyperlinkFormat));
+    }
+    else
+    {
+        // Opening OSC8 sequence
+        const auto hyperlinkFormat = til::u16u8(L"\x1b]8;id=%d;" + uri + L"\x9c");

```

Comment:
Why not just... give both the `id` and `uri` parameter to `_WriteFormattedString` and let it put the two in the string instead of running a concat to the format first?

Reply:
FYI: concatenating user data into a format string is _very dangerous(!)_. Imagine if somebody's hyperlink contained a `%s` or `%x` -- they'd pretty much have an arbitrary memory read for data left on the stack after the format args :smile:

Reply:
I was following the way `_ChangeTitle` did it (in the same file) - I guess Dustin's reply explains why it was like that

Reply:
_So_ @zadjii-msft and I tried to get `_WriteFormattedString` to work with this, but it did not seem to want to cooperate. Will post a follow-up regarding what we tried and what we found - for now we get around the issue by formatting it ourselves and simply calling `_Write`